### PR TITLE
Phase 4 of XLA int8 convolution on CUDA

### PR DESCRIPTION
**Pull request by yongfeng-nv**
This is a breakdown of previous PR (https://github.com/tensorflow/tensorflow/pull/29158) per @timshen91 's suggestion.
PLEASE NOTE: This PR DOES depend on (https://github.com/tensorflow/tensorflow/pull/30761, https://github.com/tensorflow/tensorflow/pull/30762,  and https://github.com/tensorflow/tensorflow/pull/30771).

1. Change cudnn_pad_for_tensor_cores to cudnn_pad_for_convolutions and add padding for integer convolution to it.
2. Remove checks for same types in convolution shape inference to allow mixing int8 activation with float bias.
3. Rename InitializeFloatBuffer to InitializeBuffer and add S8 support to it.
4. Add and modify tests for cudnn_pad_for_convolutions_test, cudnn_conv_rewriter_test.cc, and cudnn_fused_conv_rewriter_test.cc.

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/cudnn_pad_for_convolutions.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace xla {
+namespace gpu {
+
+// Two zero-paddings for CuDNN thunking are done in this transform: padding for
+// tensor cores and padding for integer convolutions.  This transform also
+// add slice instruction to remove unnecessary output features.
+class CudnnPadForConvolutions : public HloModulePass {
+ public:
+  explicit CudnnPadForConvolutions(bool is_volta_or_later)
```
Comment (yongfeng-nv) [inline tensorflow/compiler/xla/service/gpu/cudnn_pad_for_convolutions.h]: One review comment from the previous PR is to let CudnnPadForConvolutions take StreamExecutor*. 
It makes HLO unit test difficult.  I am not sure whether it is possible for a HLO unit test to pass a StreamExecutor to construct a CudnnPadForConvolutions object.  Please see a use case in cudnn_pad_for_convolutions_test.cc @ line 46.

Comment (timshen91) [reply tensorflow/compiler/xla/service/gpu/cudnn_pad_for_convolutions.h]: Regarding your comment in the old PR, take a look at gemm_rewrite_test.cc and `default_executor()`. Then it should be easy for `Run()` to take a `StreamExecutor*`.

File: tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter_test.cc [hunk 2]
```diff
@@ -631,6 +642,46 @@ TEST_F(CudnnConvRewriterTest, BackwardInputConvolveConstantFilter) {
...
+
+    ENTRY Test {
+      input = s8[1,2,3,3] parameter(0)
+      filter = s8[3,3,2,5] parameter(1)
+
+      ROOT conv = s8[1,5,3,3] convolution(input, filter), window={size=3x3 pad=1_1x1_1}, dim_labels=bf01_01io->bf01, feature_group_count=1
+    })");
+  TF_ASSERT_OK_AND_ASSIGN(auto m, ParseAndReturnVerifiedModule(module_str));
+
+  //   ASSERT_FALSE(CudnnConvRewriter().Run(m.get()).ValueOrDie());
```
Comment (timshen91) [inline tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter_test.cc]: I think it's sufficient to `ASSERT_FALSE(CudnnConvRewriter().Run(m.get()).ok())`.

File: tensorflow/compiler/xla/service/gpu/cudnn_pad_for_convolutions.h (hunk 1 above)
Comment (yongfeng-nv) [reply tensorflow/compiler/xla/service/gpu/cudnn_pad_for_convolutions.h]: The default_executor has both major and minor versions set to -1.  I don't know how to set versions to pretend Volta.  The tests copied from cudnn_conv_pad_for_tensor_cores_test however are only for Volta.

File: tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter_test.cc (hunk 2 above)
Comment (yongfeng-nv) [reply tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter_test.cc]: I tried the recommended change.  Although I could see the expected error message in the test log, the test didn't finish properly.

[ RUN      ] CudnnConvRewriterTest.ForwardIntegerOutputConvolution
2019-07-18 05:59:31.063159: I tensorflow/compiler/xla/service/platform_util.cc:78] platform Host present but no XLA compiler available: could not find registered compiler for platform Host -- check target linkage (hint: try linking in tensorflow/compiler/jit:xla_cpu_jit)
2019-07-18 05:59:31.063200: I tensorflow/compiler/xla/service/platform_util.cc:78] platform CUDA present but no XLA compiler available: could not find registered compiler for platform CUDA -- check target linkage (hint: try linking in tensorflow/compiler/jit:xla_gpu_jit)
2019-07-18 05:59:31.063232: I tensorflow/compiler/xla/service/platform_util.cc:78] platform Host present but no XLA compiler available: could not find registered compiler for platform Host -- check target linkage (hint: try linking in tensorflow/compiler/jit:xla_cpu_jit)
2019-07-18 05:59:31.063246: I tensorflow/compiler/xla/service/platform_util.cc:78] platform CUDA present but no XLA compiler available: could not find registered compiler for platform CUDA -- check target linkage (hint: try linking in tensorflow/compiler/jit:xla_gpu_jit)
2019-07-18 05:59:31.068303: F tensorflow/stream_executor/lib/statusor.cc:34] Attempting to fetch value instead of handling error Unimplemented: The convolution instruction with integer inputs only allows float outputs.  Insert a clamp instruction with range [-128, 127) followed by a convert instruction after the convolution instruction for int8 outputs.
external/bazel_tools/tools/test/test-setup.sh: line 310:   549 Aborted                 (core dumped) "${TEST_PATH}" "$@" 2>&1

Comment (timshen91) [reply tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter_test.cc]: Have you tried linking `//tensorflow/compiler/jit:xla_gpu_jit` to the test target?

File: tensorflow/compiler/xla/service/gpu/cudnn_pad_for_convolutions.h (hunk 1 above)
Comment (timshen91) [reply tensorflow/compiler/xla/service/gpu/cudnn_pad_for_convolutions.h]: Hmm let's not dig into this too deeply, as it's not a big deal. Keeping the bool is good.

File: tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter_test.cc (hunk 2 above)
Comment (yongfeng-nv) [reply tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter_test.cc]: What do you mean linking //tensorflow/compiler/jit:xla_gpu_jit to the test target?  How to do that?  The test doesn't specify dependency on xla_gpu_jit.

Is it because I use ().ValueOrDie() instead of passing status back to the call stack?

Comment (timshen91) [reply tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter_test.cc]: XLA defines several Bazel targets as "plugins". They are cc_libraries that the final binary (e.g. a test target) may choose to depend on or not. For GPU plugin, if the test doesn't depend on it, all GPU related code is not linked into the final binary. The error message is saying sort of that.

I think the word "linking" is confusing, but it means to add //tensorflow/compiler/jit:xla_gpu_jit to the test target as a dependency.

Comment (yongfeng-nv) [reply tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter_test.cc]: Yes, "linking //tensorflow/compiler/jit:xla_gpu_jit" makes the test working.  I have made another commit for the changes.

I saw the message about "linking" in the log and considered it information message.  Is GPU plugin required by error handling, but not by regular HLO pattern matching tests.

**Comment by gbaned** [comment]
@yongfeng-nv Could you please resolve the conflicts? Thanks!

File: tensorflow/compiler/xla/service/gpu/buffer_comparator.cc [hunk 3]
```diff
@@ -405,11 +405,13 @@ StatusOr<bool> HostCompare(se::Stream* stream, se::DeviceMemoryBase lhs,
...
   const auto canonicalize = [](ComparisonType a) -> ComparisonType {
     if (std::is_same<ElementType, Eigen::half>::value && a) {
-      constexpr ComparisonType kMaxFp16Value = 65505.;
+      constexpr ComparisonType kMaxFp16Value =
+          std::is_same<ElementType, Eigen::half>::value ? 65505. : 0;
       if (std::isnan(a)) {
         return a;
       }
-      return std::max(-kMaxFp16Value, std::min(a, kMaxFp16Value));
+      return std::max(static_cast<ComparisonType>(-kMaxFp16Value),
```
Comment (yongfeng-nv) [inline tensorflow/compiler/xla/service/gpu/buffer_comparator.cc]: This change is needed, because 65505 can't fit in int8.  Refactoring this lambda to a set of template/overloaded helper functions is another solution.

File: tensorflow/compiler/xla/service/shape_inference.cc [hunk 4]
```diff
@@ -1598,6 +1598,11 @@ ShapeInference::InferDegenerateDimensionBroadcastShape(HloOpcode operation,
         batch_group_count, feature_group_count);
   }
 
+  if (!ShapeUtil::SameElementTypeIgnoringFpPrecision(lhs, rhs)) {
```
Comment (yongfeng-nv) [inline tensorflow/compiler/xla/service/shape_inference.cc]: Revert this change from the previous commit.

File: tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter_test.cc [hunk 5]
```diff
@@ -40,30 +41,16 @@ class CudnnFusedConvRewriterTest : public HloTestBase {
...
-        alpha_side_input_scalar = "0.899994934";
-        elementwise_type = string(type);
-      }
-      string hlo_resolved_string = absl::StrReplaceAll(
-          hlo_string, {{"INPUT_TYPE", type},
-                       {"ELEMENTWISE_TYPE", elementwise_type},
-                       {"ALPHA_CONV_SCALAR", alpha_conv_scalar},
-                       {"ALPHA_SIDE_INPUT_SCALAR", alpha_side_input_scalar}});
-      string optimized_hlo_string = GetOptimizedHlo(hlo_resolved_string);
+    for (absl::string_view type : {"f16", "f32", "f64"}) {
```
Comment (yongfeng-nv) [inline tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter_test.cc]: Revert the changes on floating-point tests, as it is very difficult to use them to test int8 convolutions.  We created more int8 specific tests.

File: tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter_test.cc [hunk 6]
```diff
@@ -432,6 +441,207 @@ TEST_F(CudnnFusedConvRewriterTest, TestFusedConvClamp) {
...
+      upper = f32[] constant(127)
+      uppers = f32[1,3,3,64] broadcast(upper), dimensions={}
+
+      clamp = f32[1,3,3,64] clamp(lowers, relu, uppers)
+
+      ROOT convert = s8[1,3,3,64] convert(clamp) 
+    })");
+}
+
+TEST_F(CudnnFusedConvRewriterTest, TestFusedConvWithScaledInt8SideInputBiasInt8ToFloat) {
```
Comment (yongfeng-nv) [inline tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter_test.cc]: No need to have extra check for convert<float>(int8_side_input) in the result HLO, because if it is absorbed by mistake, CuDNN won't run properly and fail the test.

**Comment by gbaned** [comment]
@yongfeng-nv Could you please resolve the conflicts? Thanks!

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

File: tensorflow/compiler/xla/service/gpu/cudnn_conv_algorithm_picker.cc [hunk 7]
```diff
@@ -271,10 +271,10 @@ StatusOr<AutotuneResult> CudnnConvAlgorithmPicker::PickBestAlgorithmNoCache(
...
   int64 rng_state = 0;
 
-  const auto initialize_buffer = [stream, &result_shape,
-                                  &rng_state](DeviceMemoryBase buffer) {
-    InitializeFloatBuffer(stream, result_shape.element_type(), &rng_state,
-                          buffer);
+  const auto initialize_buffer = [&stream, &rng_state](
+                                     DeviceMemoryBase buffer,
+                                     const Shape& buffer_shape) {
+    InitializeBuffer(stream, buffer_shape.element_type(), &rng_state, buffer);
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/cudnn_conv_algorithm_picker.cc]: What changed here? Isn't buffer_shape.element_type() the same as result_shape.element_type() ?

Comment (yongfeng-nv) [reply tensorflow/compiler/xla/service/gpu/cudnn_conv_algorithm_picker.cc]: initialize_buffer has two call sites down here.  Before this change, both took result_shape.  Now, since one takes operand->shape() and the other takes result_shape, we need to move buffer_shape from from captures to arguments for the lambda.

**Comment by gbaned** [comment]
@yongfeng-nv Could you please resolve the conflicts? Thanks!

**Comment by gbaned** [comment]
@yongfeng-nv  Still, conflicts appearing. Could you please resolve those? Thanks!

**Comment by yongfeng-nv** [comment]
Close this PR, as its change has been moved to https://github.com/tensorflow/tensorflow/pull/30771 and https://github.com/tensorflow/tensorflow/pull/31988.
